### PR TITLE
185 posts

### LINKED ISSUE #185: Posts
**Current state**
There is no CRUD or front-end for posts within the application

**Expected state**
There is a CRUD and front-end for posts within the application

**Necessities**
#187
#189

**Branch**
[185-posts](/AjobK/Seaqull/tree/185-posts)

### LINKED ISSUE #187: Posts: Create post
Follows #186 

**Current state**
User is unable to create posts

**Expected state**
User is able to create posts without styling

**Necessities**
None

**TO DO**
- [x] Front-end editor with draftjs
- [x] Front-end API call with mobx
- [x] Front-end styling
- [x] Back-end controller
- [x] Back-end resource (Routes)

**Branch**
[187-create-post](/AjobK/Seaqull/tree/187-create-post)

### LINKED ISSUE #189: Posts: Go to specific post
Follows #185 

**Current state**
User is unable to go to specific posts on the homepage

**Expected state**
User is able to go to specific posts on the homepage with UUID url

**Necessities**
#187 

**TO DO**
- [x] Front-end view post by UUID
- [x] Back-end UUID for posts

**Branch**
[189-view-post](/AjobK/Seaqull/tree/189-view-post)

### PR DESCRIPTION
**Pull request by AjobK**
Closes #185 
Closes #187 
Closes #189 

 **What is it supposed to do**
With this PR the user is supposed to do the following things:
- Create a post (DraftJS)
- Navigate to posts (Through profile and/or homepage)
- Show owned posts on anyones profile

 **How can we test your branch**
1. Go to 185-posts
2. Try creating a post as a logged in user
3. Try navigating to posts through a profile or homepage
4. Review code

 **What broweser have been checked**
- [x] Chrome
- [x] Firefox
- [x] Opera
- [ ] Internet Explorer
- [ ] Edge
- [ ] Safari

### CODE REVIEW

**Comment by jerohero** [comment]
image
Line-height makes Korean text look weird (probably not important at all)

**Comment by ryankroon00** [comment]
allemaal ; broski

File: backend/app/dao/postDao.ts [hunk 1]
```diff
@@ -1,5 +1,6 @@
 import DatabaseConnector from '../util/databaseConnector'
 import { Post } from '../entity/post'
+import Profile from '../entity/profile';
```
Comment (jerohero) [inline backend/app/dao/postDao.ts]: ;

File: backend/app/dao/titleDao.ts [hunk 2]
```diff
@@ -8,7 +8,7 @@ class TitleDAO {
         return title
     }
     public async getTitleByUserId(id: number): Promise<Title> {
-        const repository = await DatabaseConnector.getRepository('User');
+        const repository = await DatabaseConnector.getRepository('Profile');
```
Comment (jerohero) [inline backend/app/dao/titleDao.ts]: ;

File: frontend/src/pages/post/index.js [hunk 3]
```diff
@@ -51,7 +52,7 @@ class Post extends App {
             }
 
             try {
-                postItem.innerText = convertFromRaw(JSON.parse(this.data[i].title)).getPlainText()
+                // postItem.innerText = convertFromRaw(JSON.parse(this.data[i].title)).getPlainText()
```
Comment (jerohero) [inline frontend/src/pages/post/index.js]: Comment

File: backend/app/service/postService.ts [hunk 4]
```diff
@@ -2,6 +2,11 @@ import { Request, Response } from 'express'
 import PostDAO from '../dao/postDao'
 import post from '../entity/post'
 import { v4 as uuidv4 } from 'uuid'
+import { decode } from 'jsonwebtoken';
+import UserDAO from '../dao/userDao'
+import UserDao from '../dao/userDao';
+const jwt = require('jsonwebtoken');
+require('dotenv').config()
```
Comment (jerohero) [inline backend/app/service/postService.ts:8]: ;

File: backend/app/service/postService.ts [hunk 5]
```diff
@@ -18,9 +23,19 @@ class PostService {
...
     public getPostByPath = async (req: Request, res: Response): Promise<any> => {
         const foundPost = await this.dao.getPostByPath(req.params.path)
+        const { JWT_SECRET } = process.env
+
+        let decodedId = -1;
+
+        try {
+            let decodedToken = jwt.verify(req.cookies.token, JWT_SECRET);
+            let user = await new UserDao().getUserByUsername(decodedToken.username);
+            decodedId = user.id;
```
Comment (jerohero) [inline backend/app/service/postService.ts]: ;

**Review by jerohero — APPROVED** [review]
I approve

**Comment by jerohero** [comment]
chrome_SVPXdi3WsV
Got this error after creating a post

File: README.md [hunk 6]
```diff
@@ -1,10 +1,21 @@
 # [Seaqull](http://www.seaqull.com/) [![npm version](https://img.shields.io/npm/v/react.svg?style=flat)] [![PRs Welcome](https://img.shields.io/badge/PRs-welcome-brightgreen.svg)]
 
-Seaqull is a project focused on improving the accessibility to the world of technology
+#### Commands front end
```
Comment (ryankroon00) [inline README.md:4]: allemaal . ,s

**Comment by AjobK** [comment]
> chrome_SVPXdi3WsV
> Got this error after creating a post

This is because of the homepage which is outside of the scope :-)

File: backend/app/service/postService.ts [hunk 7]
```diff
@@ -48,9 +76,35 @@ class PostService {
         newPost.published_at = new Date()
         newPost.created_at = new Date()
 
-        await this.dao.createPost(newPost)
+        const decodedToken = jwt.verify(req.cookies.token, JWT_SECRET)
+        if(decodedToken.username != null){
```
Comment (ryankroon00) [inline backend/app/service/postService.ts]: spacing is anders dan lijn 61

File: backend/app/service/postService.ts [hunk 8]
```diff
@@ -48,9 +76,35 @@ class PostService {
...
+        const { JWT_SECRET } = process.env
+
+        post.title = req.body.title
+        post.description = req.body.description
+        post.content = req.body.content
+
+        const decodedToken = jwt.verify(req.cookies.token, JWT_SECRET)
+        if(post.profile.display_name != decodedToken.username) return res.status(405).json({ 'error': 'Not allowed' })
+
+        if(!post) return res.status(404).json({ 'message': 'post not found' })
```
Comment (ryankroon00) [inline backend/app/service/postService.ts]: spatie

File: backend/app/service/postService.ts [hunk 9]
```diff
@@ -48,9 +76,35 @@ class PostService {
...
+    public updatePost = async (req: Request, res: Response): Promise<any> => {
+        const post = await this.dao.getPostByPath(req.params.path)
+        const { JWT_SECRET } = process.env
+
+        post.title = req.body.title
+        post.description = req.body.description
+        post.content = req.body.content
+
+        const decodedToken = jwt.verify(req.cookies.token, JWT_SECRET)
+        if(post.profile.display_name != decodedToken.username) return res.status(405).json({ 'error': 'Not allowed' })
```
Comment (ryankroon00) [inline backend/app/service/postService.ts]: spatie

File: backend/app/service/postService.ts [hunk 10]
```diff
@@ -48,9 +76,35 @@ class PostService {
...
+
+        if(!post) return res.status(404).json({ 'message': 'post not found' })
+        const updatedPost = await this.dao.updatePost(post)
+
+        if(!updatedPost) return res.status(404).json({ 'message': 'post not updated' })
```
Comment (ryankroon00) [inline backend/app/service/postService.ts]: spatie

File: backend/app/service/profileService.ts [hunk 11]
```diff
@@ -31,34 +32,37 @@ class ProfileService {
...
+            decodedToken = jwt.verify(token, process.env.JWT_SECRET)
+        } catch (e) { }
 
         // getting username from parameters
-        let recievedUsername = req.params.username
+        let receivedUsername = req.params.username
 
         // check if username is precent in the paramenters
-        if(!recievedUsername){
+        if(!receivedUsername){
```
Comment (ryankroon00) [inline backend/app/service/profileService.ts]: spatie

File: backend/app/service/profileService.ts [hunk 12]
```diff
@@ -31,34 +32,37 @@ class ProfileService {
...
             // check if a username is present in payload
-            if(decodedToken.username){
+            if(decodedToken && decodedToken.username){
```
Comment (ryankroon00) [inline backend/app/service/profileService.ts]: spatie

**Review by jerohero — APPROVED** [review]

File: backend/app/service/postService.ts [hunk 13]
```diff
@@ -2,42 +2,70 @@ import { Request, Response } from 'express'
...
-        const amount = 7;
+        const amount = 7
 
-        if(req.query.page == null) {
+        if (req.query.page == null) {
             posts = await this.dao.getPosts('0', amount)
         } else {
             posts = await this.dao.getPosts(String(req.query.page), amount)
-            if(posts.length == 0 ){
+            if (posts.length == 0 ){
```
Comment (ryankroon00) [inline backend/app/service/postService.ts]: ) { aub

**Review by ryankroon00 — CHANGES_REQUESTED** [review]

**Review by ryankroon00 — APPROVED** [review]

**Pull request merged**